### ISSUE TITLE
Dynamic dependency class __call__ parameter which can work with different routes

### ISSUE THREAD

**Issue by killerdejavu** [tiangolo/fastapi#2575]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import APIRouter, Depends

router: APIRouter = APIRouter()

class User:
    id: int
    name: str

class Book:
    id: int
    created_by_id: int

class Game:
    id:int
    created_by_id: int

class GetObject:
    def __init__(self, model):
        self.model = model

    async def __call__(self, id):
        return self.model.get(id=id)

async def get_current_user():
    return User()

class DefaultAuthorization:
    def __init__(self, model):
        self.model = model

    # next line is where i need help 
    def __call__(self, source_object=Depends(GetObject(**get_model_class_here**)), current_user=Depends(get_current_user)):
        assert source_object.created_by_id == current_user.id


@router.patch("/book/{id}", dependencies=[Depends(DefaultAuthorization(Book))])
async def update_book(book=Depends(GetObject(Book))):
    # update book
    pass

@router.patch("/game/{id}", dependencies=[Depends(DefaultAuthorization(Game))])
async def update_game(book=Depends(GetObject(Game))):
    # update game
    pass
```

### Description

There are two problems / questions i am facing
1. I have written a `DefaultAuthorization` so that it can check that the current logged in user is the same who created that resources (Game, Book) object. I want to use the same class in both the places and dont want to write two different class, as the core logic is still the same
2. I want to write it in a way that `Depends(GetObject(Game)` is called only once during the `DefaultAuthorization` dependency resolution and that I can get it from cached when I do `book=Depends(GetObject(Book)`

How do i write my `DefaultAuthorization` `__call__` method so that it can get the model reference dynamically `source_object=Depends(GetObject(**get_model_class_here**))` so that I can get correct object according to the api call. 

### Environment

* OS: [e.g. Linux / Windows / macOS]: MacOS
* FastAPI Version: 0.54.2
* Python version: 3.7.8

**Comment by falkben** [comment]
I don't have the best solution to your problem at the moment, but I'll get to that later... 

First, just wanted to point out that you are passing into `Depends` instances of the classes `GetObject` and `DefaultAuthorization`, which are not callables themselves.  If you read through the docs on Dependencies again, you will see that what needs to be passed into `Depends` is a callable, not an instance.

So it should be something like:

```python
@router.patch("/game/{id}", dependencies=[Depends(DefaultAuthorizationGame)])
async def update_game(book=Depends(GetGame)):
    # update game
    pass
```

Where you would need create new classes `GetGame` and `DefaultAuthorizationGame` specific to the Game model. Anyways, you don't invoke it when passing it into Depends, you pass the callable. Hopefully that makes some sense.

To your specific question about authorization, this is not a trivial problem. Perhaps there's some way to use scopes for this... 

I recently learned about [oso](https://www.osohq.com/), but haven't used it. oso solves this issue by creating a "model" that describes what authorization is allowed.  Would solve your issue I think, but I don't think it's trivial to set up, either. Would love to hear how it goes if you do use it.

**Comment by killerdejavu** [comment]
@falkben Thanks for your response! On the below thing you said

> First, just wanted to point out that you are passing into `Depends` instances of the classes `GetObject` and `DefaultAuthorization`, which are not callables themselves. If you read through the docs on Dependencies again, you will see that what needs to be passed into `Depends` is a callable, not an instance.

They both are callables. If you see, both of them implement the `def __call__` So they work well with Depends. Actually the docs nearly  the same example which I have - https://fastapi.tiangolo.com/advanced/advanced-dependencies/#a-callable-instance It just implements the `__call__` method on the instance, which makes it callable. 

Will look at oso, thanks for recommending that!

**Comment by falkben** [comment]
Please look again at that example in the docs. Specifically in the `Depends` call.

**Comment by killerdejavu** [comment]
> Please look again at that example in the docs. Specifically in the `Depends` call.

Isn't it the same 🤔 Or am I missing something 

```
class FixedContentQueryChecker:
    def __init__(self, fixed_content: str):
        self.fixed_content = fixed_content

    def __call__(self, q: str = ""):
        if q:
            return self.fixed_content in q
        return False


checker = FixedContentQueryChecker("bar")

async def read_query_check(fixed_content_included: bool = Depends(checker)):
```

will be same as 

```
class FixedContentQueryChecker:
    def __init__(self, fixed_content: str):
        self.fixed_content = fixed_content

    def __call__(self, q: str = ""):
        if q:
            return self.fixed_content in q
        return False


async def read_query_check(fixed_content_included: bool = Depends(FixedContentQueryChecker("bar"))):
```

For my case also, I can write it as 
```

authorization_callable = DefaultAuthorization(Book) # instance returned will be callable ass it would have __call__
get_object_callable = GetObject(Book) # instance returned will be callable ass it would have __call__

@router.patch("/book/{id}", dependencies=[Depends(authorization_callable)])
async def update_book(book=Depends(get_object_callable)):
    # update book
    pass
```

which is the same (difference is they pass a string, and I pass a reference to class). Also my question was exactly how to not implement route specific dependencies like `DefaultAuthorizationGame` and `GetGame` as the underlying logic remains that same, so find a way to just pass reference to Game class as a parameter.

**Comment by falkben** [comment]
Huh, I thought there was something during signature inspection that prevented you from invoking a function inside the Depends call. Or something related to mutable parameters in the signature... Anyways, appears I was incorrect about that. Sorry for the confusion.

In this case, I suppose you could sequence multiple dependencies together in order to use a common authorization class. On any routes that are "book" classes, you could prepend a `set_book_model` dependency to your list of dependencies. It would look something like:

```python
def set_book_model(request: Request):
    request.state.model = Book
```

So then you dependency list would be come something like

```python
dependencies=[Depends(set_book_model), Depends(DefaultAuthorization)]
```

For the game routes, you would just use a `set_game_model` dependency.

Then in your `DefaultAuthorization` dependency, you could grab that model out of the request object.

I think this could work (you might have to restructure your DefaultAuthorization dependency so that it can access the request object).  However, this feels a bit like an anti-pattern to me, just in order to reuse the authorization code, which is almost certainly going to diverge based on different models anyways...

Alternatively, maybe you can do something with class inheritance, where the BookAuthorization inherits from DefaultAuthorization, but defines a `self.model = Book`. But the implementation of `__call__` might be able to be left unchanged between the different models?

**Comment by killerdejavu** [comment]
Yes, I forgot it can be stored in state. That looks like decent solution for now. Also to make it generic you can do

```
def set_model(model):
    def set_(request: Request):
        request.state.model = model
    return set_
```

And then you can use it in depends like 
```
dependencies=[Depends(set_model(Book)), Depends(DefaultAuthorization)]
```
or 
```
dependencies=[Depends(set_model(Game)), Depends(DefaultAuthorization)]
```

Closing this issue, I have ok solution for now.

**Comment by tiangolo** [comment]
Thanks for the help here @falkben ! 👏  🙇 

Thanks for reporting back and closing the issue @killerdejavu  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
